Noise overhaul (Add more noise types, noise color ramp, replace OpenSimplexNoise with FastNoiseLite)

Geometror opened the pull request:
Co-authored-by: Cory Petkovsek <tinmanjuggernaut@users.noreply.github.com>

This PR aims to restructure and improve how noise is generated in Godot. Based on the amazing work by @tinmanjuggernaut in #35144.
Alternative to/supercedes : #35144
Implements: godotengine/godot-proposals#1493

![Screenshot 2022-01-11 202003](https://user-images.githubusercontent.com/50084500/149134772-1a986f45-e160-4aab-a666-4d4a2eb342e1.png)

Main features:
- Implements FastNoiseLite
   -    addition: color ramp
- Adds a property to control the blend skirt width for seamless noise textures which are generated by a slightly modified implementation of the algorithm used in #35144
(Optimized image creation (choosing the fitting image format))
- Adds a Noise interface (Abstract Noise resource with virtual methods `get_noise_1/2/3d()`, `get_image()` and `get_seamless_image()`)
- General restructuring (files moved/comment style/…)
   -    Move offset from NoiseTexture to the different Noise resources so that each resource inheriting from Noise can decide whether to provide this property.

Some demos:
![Animation4](https://user-images.githubusercontent.com/50084500/149133289-8017436b-073e-4e83-852a-92262cd18868.gif)
![Animation3](https://user-images.githubusercontent.com/50084500/149132951-d4cbd385-7a65-4cc1-b458-cb702858be00.gif)
![grafik](https://user-images.githubusercontent.com/50084500/149133430-d01e1c77-55ed-40d1-b3ae-4c6adf8372a8.png)

[Optional TODOs]:
- Some changes to the noise interface (make `get_noise_...d()` part of interface?)
- Allow float texture formats (FORMAT_RF, FORMAT_RGBAF)
- ~~Add NoiseTexture3D (useful for fog volumes)~~ For another PR

Feedback is always appreciated :)

clayjohn commented:
NoiseTexture3D is probably best done in another PR. 

Overall, I am torn about adding FastNoise ([despite the fact that I was the one who originally proposed it](https://github.com/godotengine/godot/issues/22337)). 

I love the idea of having a noise library integrated into Godot. It is great for procedural systems, generative art, and for making cool VFX. 

At the same time, FastNoise exposes a ton of detailed handles which makes it very complex and daunting for end-users. One of Godot's goals is to keep things simple and easy to comprehend for the average user. As a result we often sacrifice control over minor details in order to keep things comprehensible. 

To me, FastNoise sits on the border of value and complexity. On the one hand, I think it is valuable, but on the other hand it introduces significant complexity (and another 3rd party dependency). 

After typing all this out, I am finding myself more supportive of merging than not. Especially if we could somehow replace OpenSimplexNoise entirely, then the benefits would more clearly outweigh the added complexity.

Zireael07 commented:
If not merged into core, it should be an officially endorsed plugin (like what was proposed for VehicleBody some time back)

Even if it *does* add some complexity, it gives users a LOT a LOT of power and the ability to easily achieve/port effects seen in e.g. shadertoy.

mrjustaguy commented:
I don't think Fast Noise is that complex, anyone that plays around with it a little will be able to quickly find their way around and it's not something where you can unintentionally be creating possible artifacts that wouldn't be clearly visible in the preview and that occur only in specific situations.

tinmanjuggernaut commented:
@clayjohn 
> FastNoise exposes a ton of detailed handles which makes it very complex and daunting for end-users. One of Godot's goals is to keep things simple and easy to comprehend for the average user. As a result we often sacrifice control over minor details in order to keep things comprehensible.

GDScript is complex and daunting, as are C# and C++. Yet they are all natively supported languages. The capabilities do not detract from the stated goal. 

I would agree with you if the noise system was like, say Zylann's original voxel system where you had to get several variables set up properly to get it working. Otherwise it gave zero results. (It was fine later once we set it up with defaults and documentation.) But that's not the case in my PR, and I assume about this one. They work with default values as OpenSimplexNoise does.

Having a noise library with a bunch of options that all provide visual feedback does not require the user to understand and comprehend what everything does. They can push a bunch of buttons and get direct visual feedback as to how that changes the end result. That's far more user-friendly than half the things in the engine right now. 

e.g. Looking through the VisualServer, I can see it would not be merged if it required `simplicity` and `comprehensibility` for the average user. Same thing with Environment, or your PhysicalSky. Does the average user know what Mie, Turbidity, or Rayleigh are? A better metric is `useability` for the average user, which both noise PRs provide.

> Especially if we could somehow replace OpenSimplexNoise entirely, then the benefits would more clearly outweigh the added complexity.

Perlin and now simplex ([The patent expired 2022-01-08!](https://patents.google.com/patent/US6867776B2/en)) are suitable replacements for OSN. However due to class names and the like, you'll have to make a breaking decision to deprecate and remove OSN. 4.0 is a good time to do that.

> Overall, I am torn about adding FastNoise ([despite the fact that I was the one who originally proposed it]
...
> After typing all this out, I am finding myself more supportive of merging than not. 

Yes, so now at least three people have put in a lot of work to make your request possible. Us two and Zylann. And we haven't even received the courtesy of rejection, or rejection for reasons. Silence after completing the request is neither community-building nor contributor-encouraging.

So how about this. You were the original champion. I propose you take up that again and get a firm decision from the dev team on these questions:
* Do you want to support multiple noise libraries in any capacity? Right now OSN _is_ NoiseTexture. There is no abstract concept of Noise and NoiseTexture that can allow something other than OSN. There is zero support for 3rd party modules or plugins. To get a module to work, a gamedev has to hack up the engine. Case in point: https://github.com/tinmanjuggernaut/godot_fastnoise_simd#preparation
* Do you want fastnoise in core or only as a plugin or module?
* Do you want to replace OpenSimplexNoise since it is redundant with Fastnoise?

Then we could insert smaller, more digestible PRs such as the following.
1. Noise - abstract class to separate Noise from OpenSimplexNoise
2. FastNoise/Lite - to add to, or replace OSN
3. NoiseTexture3D

If the first is accepted, which allows support for modules and plugins, there are at least three additional noise modules available to users: fastnoise, fastnoise_simd, and fastnoise_lite, and room to convert them to plugins. OSN also has a new version with new algorithms one might be interested in implementing.

However, I for one am not putting out more effort on this until there is a decision and clear direction from the dev team.

albinaask commented:
Really love the capabilities this PR seem to have. Incredible work, It looks epic! I really feel like it's something Godot has lacked, and I definitely think it should be in the core! 

@clayjohn. One possibility to make it easier for new users and simpler use cases would be to expose two variants of the noise as different resource types. One "simple" with a lot of the functionality hidden and set to default values basically filling the place of the old OpenSimplexNoise, but using the same code base and thus saving 2000 lines of code. Then expose a "flexible" variant with the full feature set and a lot of options for more demanding use cases.

Calinou commented:
> One possibility to make it easier for new users and simpler use cases would be to expose two variants of the noise as different resource types. One "simple" with a lot of the functionality hidden and set to default values basically filling the place of the old OpenSimplexNoise, but using the same code base and thus saving 2000 lines of code. Then expose a "flexible" variant with the full feature set and a lot of options for more demanding use cases.

This sounds like a good idea on paper, but one issue is that to convert from "simple" to "advanced" noise without recreating the resource in the editor, it will require a bespoke conversion tool. It is possible to make such a tool (see BaseMaterial3D's "Convert to ShaderMaterial" functionality), but it isn't easy to discover for users.

Geometror commented:
Updated this PR to fix some issues with properties, clean them up and find good default values which come close to OpenSimplexNoise defaults:

![grafik](https://user-images.githubusercontent.com/50084500/149430971-6a612ce8-1cee-4a74-9e76-db8c5f354f89.png)

@clayjohn I made sure that all properties that have no effect for a specific (fractal) noise type or with a specific combination are hidden, making it quite easy to use despite having lots of detailed handles. Give it a try :)

OpenSimplexNoise is now removed.
As discussed here and via chat, it makes sense to replace OpenSimplexNoise with FastNoiseLite. Other noise types can still be added via extending the Noise interface. The only feature missing in FastNoiseLite is generating 4-dimensional noise and therefore "true" seamless noise, but the new blending algorithm works very well in most cases. But for better tileable noise support (also in 3d) something like https://github.com/Xrayez/godot-anl could be used (ported to GDExtension).

Geometror commented:
@albinaask @Calinou I don't think it is necessary to introduce two resource types. As it was already said, it is not that complex to work with :) 
The `seamless_blend_skirt` property is only used when a custom noise resource does not implement seamless noise itself and relies on the fallback implementation. We could add a virtual method to the `Noise` interface to get that capability and hide the property accordingly.

clayjohn commented:
> I made sure that all properties that have no effect for a specific (fractal) noise type or with a specific combination are hidden, making it quite easy to use despite having lots of detailed handles. Give it a try :)

That definitely helps reduce the complexity! 

I'm wondering if it is possible to further reduce the complexity by exposing FastNoiseLite in a more "Godot-like" fashion. Right now a single noise node exposes every possible configuration (including FBM and Domain warping). I wonder if it would be simpler from an end user perspective using Godot-style interfaces. 

What I have in mind is exposing 4 noise classes: Noise, Cellular, Fractal, and Domain Warp. Fractal and Domain warp would themselves accept a new "noise" parameter which itself would be one of the 4 noise types. 

```
Noise {
    type = Simplex, SimplexSmooth, Perlin, Value, ValueCubic
    seed;
    frequency;
    offset;
    in_3d_space;
    ramp;
}
Cellular {
    seed;
    frequency;
    offset;
    in_3d_space;
    ramp;
    distance_function;
    jitter;
    return_type;
}
Fractal {
    noise = Noise, Cellular;
    type;
    octaves;
    lacunarity;
    gain;
    weighted_strength;
    pingpong_strength;
}
DomainWarp {
    noise = Noise, Cellular, Fractal
    type;
    amplitude;
    frequency;
}
```

From a brief look at the code, the above seems possible. Keep in mind, this is just an idea to discuss. In my mind, this would allow us to expose a relatively simple user interface, but would still allow making just as complex effects as in the original PR. It may, however, require us to work around FastNoiseLite as I see that it engrains concepts like "Fractal" pretty deeply into the concept of "noise". 

Another point for discussion, but is it useful for end users to have a choice between Simplex, SimplexSmooth, Perlin, Value, and Value cubic? The difference to end users will be quite minor and may not to even be noticeable to many. My guess is that 99% of users will be fine with Simplex and won't ever need to look at the other types (especially when using FBM where the peculiar visual artifacts of simplex are less pronounced).

clayjohn commented on modules/noise/doc_classes/FastNoiseLite.xml:
```diff
@@ -0,0 +1,208 @@
+			A lattice of points are assigned random values then interpolated based on neighboring values.
+		</constant>
+		<constant name="TYPE_VALUE_CUBIC" value="4" enum="NoiseType">
+			Similar to Value noise, but slower. Has more variance in peaks and valleys.
```
Importantly, cubic interpolation avoids discontinuities in the 2nd order derivative of the noise function. This is important when using noise as a bumpmap. Regular value noise will show sharp discontinuities when used as a normal map that cubic interpolation avoids

clayjohn replied:
```suggestion
			Similar to Value noise, but slower. Has more variance in peaks and valleys. Cubic noise can be used to avoid certain artifacts when using value noise to create a bumpmap. In general, you should always use this mode if the value noise is being used for a heightmap or bumpmap.
```

Geometror replied:
Oh, sorry, I've missed this one. Now it's added.

clayjohn commented on modules/noise/doc_classes/FastNoiseLite.xml:
```diff
@@ -0,0 +1,208 @@
+		<constant name="TYPE_CELLULAR" value="2" enum="NoiseType">
+			Cellular includes both Worley noise and Voronoi diagrams which creates various regions of the same value.
+		</constant>
+		<constant name="TYPE_SIMPLEX" value="0" enum="NoiseType">
```
An important distinguishing feature of simplex noise over other gradient noise types is that the gradients exist in a simplex lattice rather than a grid lattice. This helps avoid directional artifacts that perlin noise suffers from. It also means that it scales better with an increase in dimensions (once you hit 4 or more dimensions simplex noise becomes much faster than perlin)

albinaask commented:
> > One possibility to make it easier for new users and simpler use cases would be to expose two variants of the noise as different resource types. One "simple" with a lot of the functionality hidden and set to default values basically filling the place of the old OpenSimplexNoise, but using the same code base and thus saving 2000 lines of code. Then expose a "flexible" variant with the full feature set and a lot of options for more demanding use cases.
> 
> This sounds like a good idea on paper. One issue is that to convert from "simple" to "advanced" noise without recreating the resource in the editor, it will require implementing a bespoke resource conversion tool. It is possible to make such a tool (see BaseMaterial3D's **Convert to ShaderMaterial** functionality in the inspector toolbox), but it isn't easy to discover for users.
> 
> ~Still, I see at least one way to condense the current selection of properties exposed to the user. The `seamless` and `seamless_blend_skirt` properties can be merged in one, automatically enabling seamless mode if `seamless_blend_skirt` is greater than 0.0. Then the `seamless` property can be removed.~
> 
> **Edit:** The above is problematic for custom noise implementations (assuming OpenSimplexNoise is removed from core when this PR is merged). This is because custom noise implementations can have proper seamless support instead of relying on a blend skirt.

That wasn't really what I meant, but rather to just keep the two resources as two alternatives for different use cases, not to be convertible to one other, but it was just a quick thought ;)

ccuntd commented:
do you use https://github.com/Auburn/FastNoise2

![image](https://user-images.githubusercontent.com/85245131/149835775-3de81e3c-a433-4a34-ab3d-20462a6000b1.png)
because it's better than fastnoiselite

tinmanjuggernaut commented:
@ccuntd FN2 uses simd. I already wrote a module for FastNoise simd. https://github.com/tinmanjuggernaut/godot_fastnoise_simd 

It won't be in Godot core as long as mingw-64 is part of the cross compiling tool chain.
https://github.com/tinmanjuggernaut/godot_fastnoise_simd#supported-compilers--hardware

However if the abstract Noise module is accepted, anyone could build their own instance of FastNoise simd 1 or 2, even for all platforms.

npetrangelo commented:
Given this, I assume it's possible to add the 4D noise as an extension?
> Other noise types can still be added via extending the Noise interface. The only feature missing in FastNoiseLite is generating 4-dimensional noise and therefore "true" seamless noise, but the new blending algorithm works very well in most cases. But for better tileable noise support (also in 3d) something like https://github.com/Xrayez/godot-anl could be used (ported to GDExtension).

fire-forge commented:
This is incredible! As someone who's worked on several projects involving procedural generation, having this would create so many new possibilities. 

A few small suggestions:
* "Warp" can be removed from these property names in the inspector because all of the properties in that category have it (lots of other nodes do this).
![149133430-d01e1c77-55ed-40d1-b3ae-4c6adf8372a8](https://user-images.githubusercontent.com/67974470/152483904-7a3d953c-e1a2-4818-8287-37e6ddc43e00.png)

* "Dist Func" can be changed to "Distance Function"

Geometror commented:
Implemented the suggested changes by @fire-forge:

![grafik](https://user-images.githubusercontent.com/50084500/153693239-540f616b-a056-4c57-9f43-9ce67cc64d67.png)
(Note to reviewers: Removed the "warp" prefix by changing the prefix in the ADD_GROUP macro, I think this should be the right way, pls let me know if a different approach is preferred)

Geometror commented:
@npetrangelo Yes, this should be possible.

Any opinions on whether we should keep the noise interface as it is or make the `get_noise_Xd` functions part of it? (Maybe just get_noise_1d/-2d?)

npetrangelo commented:
@Geometror I think there should be a get_noise_3d() function because that enables the smooth evolution of a noise texture using time as the 3rd dimension.

Geometror commented:
@npetrangelo This function is already present as part of the fastnoise_lite resource, but my question was rather about which operations a custom noise resource must implement mandatorily. Having a get_nosie_3d method in the noise interface would mean that every custom noise generator needs to implement it, even if it makes no sense/is not possible. But on the other hand, these noise generators could just project the 3d space to a 2d plane or implement some other fallback (this behavior could also be controlled via a property), which would make the interface more versatile. This, however, raises the question of where to draw the line: Should get_nosie_4d/5d/6d also be a part of the noise interface?

fire-forge commented:
@Geometror for noise types that only support 2D, the 3D noise functions could just return the 2D noise value with z ignored. And then it should be noted in the docs which noise types do and don't support 3D noise.

I think having 4D noise would be good, especially if fastnoise lite already has it. It could be used for 3D noise + time. 5D and 6D noise seems unnecessary though, I wouldn't include them.

Geometror commented:
Rebased and moved the following methods into the noise interface:

```
virtual real_t get_noise_1d(real_t p_x) = 0;

virtual real_t get_noise_2dv(Vector2 p_v) = 0;
virtual real_t get_noise_2d(real_t p_x, real_t p_y) = 0;

virtual real_t get_noise_3dv(Vector3 p_v) = 0;
virtual real_t get_noise_3d(real_t p_x, real_t p_y, real_t p_z) = 0;
```
get_noise_4d is not (yet) part of it since FastNoiseLite unfortunately does not support 4D noise atm and I don’t like to add a function to the interface which is not fully implemented by the new default noise implementation. But keep in mind this doesn't stop other noise generators to provide this functionality, it is just not a mandatory requirement for a generator extending the noise interface.

clayjohn reviewed (changes requested):
We had a lengthy discussion about this in today's PR review meeting. The consensus was to merge this for now and then evaluate user feedback during the alphas and betas. With the new noise abstract class, it seems like FastNoise would be a great fit for a first-class GDExtension, but the GDExtension system is still relatively new, and we want to get this in the engine and get user feedback now.

Depending on user feedback, (i.e. if users find the current options too complicated) we are considering exposing a simpler type of noise only by default and then exposing the full range of noise types and parameters in an official GDExtension. 

To merge this, the documentation needs to be finished.

One other consideration is ``get_noise_4d()`` I don't like taking away a function that was available in 3.x. Is there any way that we can implement something on top of FastNoise that exposes ``get_noise_4d()``?

Geometror commented:
Completed the documentation, I am open for improvements as describing noise algorithms in an easy and understandable manner can be quite difficult :)

Unfortunately, adding a get_noise_4d() function to the interface would require either gluing some code on top of FastNoiseLite from the legacy FastNoise library (this would only be possible for simplex noise) or providing a dummy implementation which would just project the 4d space onto a 3d space and use the already present 3d function.
The first option involves quite some work since the library wasn't designed for 4d noise and the second one could be confusing for someone who expects true 4d noise, so I think we should stick to the get_noise_1d/2d/3d functions for now. Maybe 4d noise gets added to FastNoiseLite in the future.

clayjohn commented on modules/noise/doc_classes/FastNoiseLite.xml:
```diff
@@ -0,0 +1,166 @@
+	</tutorials>
+	<members>
+		<member name="cellular_distance_function" type="int" setter="set_cellular_distance_function" getter="get_cellular_distance_function" enum="FastNoiseLite.CellularDistanceFunction" default="0">
+			Determines how the distance to the nearest/second-nearest point is computed.
```
```suggestion
			Determines how the distance to the nearest/second-nearest point is computed. See [enum CellularDistanceFunction] for options.
```

clayjohn commented on modules/noise/doc_classes/FastNoiseLite.xml:
```diff
@@ -0,0 +1,166 @@
+			A [Gradient] which is used to map the luminance of each pixel to a color value.
+		</member>
+		<member name="domain_warp_amplitude" type="float" setter="set_domain_warp_amplitude" getter="get_domain_warp_amplitude" default="30.0">
+			Sets the maximum warp distance from the origin.
```
Is this in pixel-space? It should be specified what unit warp distance is measured in

Geometror replied:
It uses the "standard" unit of FastNoise, so I don't think we can specify a unit here. When generating a noise texture one unit is equal to one pixel, but when using the library directly throught the noise function, we can't call it pixel-space.

clayjohn commented on modules/noise/doc_classes/FastNoiseLite.xml:
```diff
@@ -0,0 +1,166 @@
+			If enabled, another FastNoiseLite instance is used to warp the space, resulting in a distortion of the noise.
+		</member>
+		<member name="domain_warp_fractal_gain" type="float" setter="set_domain_warp_fractal_gain" getter="get_domain_warp_fractal_gain" default="0.5">
+			Octave gain of the fractal noise which warps the space.
```
Gain increases (or decreases) the strength of each subsequent layer of noise in fractal noise. A low value places more emphasis on the lower frequency base layers, while a high value puts more emphasis on the higher frequency layers

clayjohn commented on modules/noise/doc_classes/FastNoiseLite.xml:
```diff
@@ -0,0 +1,166 @@
+			Octave gain of the fractal noise which warps the space.
+		</member>
+		<member name="domain_warp_fractal_lacunarity" type="float" setter="set_domain_warp_fractal_lacunarity" getter="get_domain_warp_fractal_lacunarity" default="6.0">
+			Octave lacunarity of the fractal noise which warps the space.
```
Lacunarity is the frequency multiplier between subsequent octaves. It makes higher octaves produce noise with finer details and a rougher appearance

clayjohn commented on modules/noise/doc_classes/FastNoiseLite.xml:
```diff
@@ -0,0 +1,166 @@
+			The method for combining octaves into a fractal which is used to warp the space. See [enum DomainWarpFractalType].
+		</member>
+		<member name="domain_warp_frequency" type="float" setter="set_domain_warp_frequency" getter="get_domain_warp_frequency" default="0.05">
+			Frequency of the noise which warps the space.
```
Low frequency results in smooth noise while high frequency results in rougher noise

clayjohn commented on modules/noise/doc_classes/FastNoiseLite.xml:
```diff
@@ -0,0 +1,166 @@
+			Sets the warp algorithm. See [enum DomainWarpType].
+		</member>
+		<member name="fractal_gain" type="float" setter="set_fractal_gain" getter="get_fractal_gain" default="0.5">
+			Octave gain for all fractal noise types.
```
Gain increases (or decreases) the strength of each subsequent layer of noise in fractal noise. A low value places more emphasis on the lower frequency base layers, while a high value puts more emphasis on the higher frequency layers

clayjohn commented on modules/noise/doc_classes/FastNoiseLite.xml:
```diff
@@ -0,0 +1,166 @@
+			Octave gain for all fractal noise types.
+		</member>
+		<member name="fractal_lacunarity" type="float" setter="set_fractal_lacunarity" getter="get_fractal_lacunarity" default="2.0">
+			Octave lacunarity for all fractal noise types.
```
Lacunarity is the frequency multiplier between subsequent octaves. It makes higher octaves produce noise with finer details and a rougher appearance

clayjohn commented on modules/noise/doc_classes/FastNoiseLite.xml:
```diff
@@ -0,0 +1,166 @@
+			Higher weighting means higher octaves have less impact if lower octaves have a large impact.
+		</member>
+		<member name="frequency" type="float" setter="set_frequency" getter="get_frequency" default="0.01">
+			The frequency for all noise types.
```
Low frequency results in smooth noise while high frequency results in rougher noise

Geometror commented:
Included the suggestions and did some additional tweaks regarding the documentation.

clayjohn reviewed (approved):
Looks good to me. Let's get this out there and start getting some feedback!

Geometror committed:
Add FastNoiseLite / general noise overhaul
- replace OpenSimplexNoise

Co-authored-by: Cory Petkovsek <tinmanjuggernaut@users.noreply.github.com>

akien-mga commented:
Thanks!

Zylann commented on modules/noise/fastnoise_lite.h:
```diff
@@ -0,0 +1,237 @@
+	real_t get_noise_2d(real_t p_x, real_t p_y) override;
+
+	real_t get_noise_3dv(Vector3 p_v) override;
+	real_t get_noise_3d(real_t p_x, real_t p_y, real_t p_z) override;
```
None of the `get_noise` functions are constant, that looks like an oversight. They can be, see my implementation https://github.com/Zylann/godot_voxel/blob/4c3777ce00473efb426362b83d3c3d5853a2ced0/util/noise/fast_noise_lite.h#L101

Geometror replied:
You're right, for FastNoiseLite they could be `const`. But then the interface methods would need to be `const` too and I am not sure if we should change it (since someone might want to add a noise generator which uses some cache structure internally or something else which requires write access in the get_noise/get_image functions). Or should we prohibit that as it might lead to bad library/API design?

Zylann replied:
It sounds like bad design to me. I care about this in particular because I use one of these in a multithreaded generator, and seeing a non-const call to a noise class from multiple threads is a bit worrying^^"

Zylann commented on modules/noise/noise.h:
```diff
@@ -0,0 +1,239 @@
+	// Virtual destructor so we can delete any Noise derived object when referenced as a Noise*.
+	virtual ~Noise() {}
+
+	virtual real_t get_noise_1d(real_t p_x) = 0;
```
Does pure virtual means extensions cannot define new noises?

tinmanjuggernaut replied:
`virtual` means referencing a `Noise*` that points to a `FastNoiseLite` object will call the FastNoiseLite implementation in the derived class rather than the implementation in `Noise` without casting.

`= 0` means there is no implementation, thus one cannot instantiate Noise().

I don't know about GDExtensions, but when I wrote this as a module this was so I could define OpenSimplexNoise, FastNoise, and FastNoiseSIMD modules.

Geometror replied:
I just noticed that the functions should actually bound with GDVIRTUAL_BIND(), but I am not exacly sure.

tinmanjuggernaut commented on modules/noise/noise.h:
```diff
@@ -0,0 +1,239 @@
+	}
+
+	template <typename T>
+	T _alpha_blend(T p_bg, T p_fg, int p_alpha) const {
```
What was the idea behind converting this and all of these functions to a template? What other types will be used here? This alpha_blend function is specifically designed for a 32-bit long int, the `l` component of the l2c union. I think there will be unexpected results with other types.

Are there any active use cases right now for generating non-32-bit noise images? Were they tested and do they work, including with the alpha blend seamless function?

Geometror replied:
I think the purpose of this was to allow for grayscale image optimizations when a noise generator creates an image with `Image::FORMAT_L8` (see the template specialization and the get_seamless_image implementation in noise.cpp). 
But looking at the code now I think there is room for improvement, which I might include in a subsequent PR. E.g. I can't remember why I didn't just overload the alpha blend function.

tinmanjuggernaut commented:
> It might be useful to have the ability to run FastNoise on a Thread instead of running on the Main Thread, so that games don't stall when used too much, and also so that Projects using Noise Materials don't take a bit to load when large noise textures are used in the project, but instead just load (even if without textures for a little bit)

Textures are generated fast enough to animate, though if you want a real dynamic noise texture it should be implemented as a shader. Otherwise if you're generating a noise texture so you can use it in a material or terrain, you can't do anything with it until the texture is generated, regardless of which thread it's rendered on. I don't think multithreaded is necessary, IMO.


@Geometror @clayjohn How about a version for 3.x? I had made a branch for both versions.

mrjustaguy commented:
I didn't mean this for dynamic noise textures. The problem is with running on main thread, it freezes the project until generation of the noise is done, and there are MANY cases where you want to generate while the rest of the project is running, without causing stutters or freezes while it's generating... 
The thing is you can tell it to generate noise textures at start up, but not actually need them until much later, so it's stupid to force long loading at start or when texture is actually needed when you can hide it away by doing it on a thread.. and in games, if you dump someone into something too early, it's very much preferable that a texture isn't loaded compared to a freeze/stutter, so yes, allowing the option to run this on a Thread would significantly help the smoothness of applications using it.

Geometror commented:
@mrjustaguy The noise generation is already offloaded to its own thread when using `NoiseTexture`. If you want to use any noise generator directly you could create a thread yourself and just call it from there (I guess, but I am not too familiar with the Godot threading API).

@tinmanjuggernaut As you said, OpenSimplexNoise needs to be kept in Godot3 since FastNoiseLite does not provide 4D noise yet. As having several noise generators in core is not desired (decided in earlier discussions), the only option would be to backport the Noise interface, which might be doable without breaking compatibility.

mrjustaguy commented:
@Geometror upon further testing I see that that specifically isn't the issue, as a larger texture (that takes longer to make) stutters before applying (seemed like texture generation to me for the texture size I used), so generation isn't the issue, it's moving it to VRAM in one go that stutters... so Texture Streaming is actually what would be needed for this to not cause a stutter/micro freeze...

tinmanjuggernaut commented:
@mrjustaguy If you are stuttering when creating a noise texture that is not displaying on screen, perhaps you have properly identified the cause. But if you're creating a texture in a material and displaying it on screen, then the engine is halting while it compiles the shader. Blocking shader compilation is a rampant problem in 3.x. It has received a bandaid in 3.5 https://github.com/godotengine/godot/pull/53411, but isn't perfect.

@Geometror I wrote the noise interface for 3.2 and 4 so we could provide for many noise types. Since the devs decided they do want noise, the interface should be added IMO.

clayjohn commented:
> @Geometror @clayjohn How about a version for 3.x? I had made a branch for both versions.
Edit: Godot 3 needs to maintain opensimplexnoise of course, but Noise and FNL can be added.

I don't have any issue with adding new noise types to 3.x so long as we can avoid any compatibility breakage.

Geometror commented:
Ok, then here is the plan: I will do a subsequent PR to this to with some improvements I had in mind, also incorporating the feedback we get in the following weeks.
Ideas/planned changes:
- Maybe bring back 4d noise and adding it to the interface [I previously said this might require too much work but I will do a bit of experimenting, it might be not that bad]
- Supporting float texture formats (FORMAT_RF, FORMAT_RGBAF)
- Some code refactoring
After we got everything right, then we could decide whether or how backporting this is viable/desired.

clayjohn commented:
Sounds good to me! :)

Zylann commented:
> It might be useful to have the ability to run FastNoise on a Thread instead of running on the Main Thread, so that games don't stall when used too much, and also so that Projects using Noise Materials don't take a bit to load when large noise textures are used in the project, but instead just load (even if without textures for a little bit)

@mrjustaguy IMO it should not be the job of the noise class to offload work to a thread. It should be done with a helper or a job system of some kind.

I also figured out that if I drop my own `FastNoiseLite` implementation for Godot's, and later move to GDExtension, the cost of calling it is going to go way up. The reason is, calls to `get_noise` from an extension have much more indirection, and cannot be inlined.
There usually one way to workaround slow calls (here pseudocode):
```
void generate_noise_series(Vector3 *input_positions, int input_count, float *output_noise)
```
However, no Variant type allows to do this. `PackedVector` types use CoW so return by reference is not possible, and passing + returning just to avoid the allocation is a bit awkward (not to mention I need to preallocate the vector somewhere in my library exclusively for Godot). So I might have to keep and prefix my version...

Another thing, more on usability side, I have this in my implementation:
![image](https://user-images.githubusercontent.com/1311555/161158596-1cd29679-d7f7-4994-9da8-4498ef3a780a.png)
My use case does not involve textures at all, so it is really useful to have a direct preview of the noise right above it. It's even fast enough to not need a thread.

mrjustaguy commented:
> @mrjustaguy The noise generation is already offloaded to its own thread when using `NoiseTexture`.

It is generated on a Thread, the issue so it seems my issue lies with sending the Textures to VRAM that cause the stutters/micro freezes.. That or Shader Compilation..

eltharionx commented:
> > @Geometror @clayjohn How about a version for 3.x? I had made a branch for both versions.
> > Edit: Godot 3 needs to maintain opensimplexnoise of course, but Noise and FNL can be added.
> 
> I don't have any issue with adding new noise types to 3.x so long as we can avoid any compatibility breakage.

This would be HUGE! Especially for us who can't really port over to 4.0.

charlesmlamb commented:
I think this is right on the edge, of what should be in Godot, add more, and it almost makes the engine into a creative tool, it's just on that maybe edge . . .

For me, it'll be easier to make effects, also some textures, etc . . It'll work esp. well in visual shaders, where one can make SOME effects, and, sort-of see what one gets, not needing to use another program, then, update Godot . .

I think, the real question is, how much Godot should be a creative program, or, just a place where one uses stuff, made in say other programs, maybe it's dumb, but, I think this is as much as Godot should have, to me, this is a 9.8 / 10 suggestion, and, makes Godot much easier, maybe, intuitive to use, it's about finding the balance, between what Godot should, be able to do, and, what is left to other programs, so on . . . This, is right on the edge, imo from the good side . . . I do wish there was a way to use simplexnoise, bec. it was more user-friendly, why not just have ' simplenoise ', and ' advancednoise ', so on, also, have gotten used to simplexnoise, I'd maybe keep both, for a while, maybe word is phase simplexnoise out, or, add bit complexity there, I think this is a good choice, for the engine, however, it's right on the edge, of what an engine should do, just 'right', etc . .

mrjustaguy commented:
Simplexnoise doesn't have the name because it's simple. Afaik it's because of https://en.wikipedia.org/wiki/Simplex

The basics here are the same, there are just some more options that you don't have to use.. having "simple" and "advanced" noise would probably just end up cluttering things up more.

aaronfranke commented on modules/noise/doc_classes/FastNoiseLite.xml:
```diff
@@ -0,0 +1,169 @@
+		<member name="cellular_return_type" type="int" setter="set_cellular_return_type" getter="get_cellular_return_type" enum="FastNoiseLite.CellularReturnType" default="0">
+			Return type from cellular noise calculations. See [enum CellularReturnType].
+		</member>
+		<member name="color_ramp" type="Gradient" setter="set_color_ramp" getter="get_color_ramp">
+			A [Gradient] which is used to map the luminance of each pixel to a color value.
```
Why isn't it just called `gradient`?

Zylann replied:
I dont know if that's related since I didnt implement this PR, but when speaking of noise, "gradient" can also relate to the derivative of noise (which in practical terms gives normals, or displacement vectors instead of a single value). This property is for the very specific use case of generating a colored image with a fixed conversion formula used exclusively in the `get_image` function. Maybe it could be called `color_gradient`?

tinmanjuggernaut replied:
Ramp is the term used in industry standard VFX tools like Adobe AfterFX and Houdini, even Blender. 

Gradient is a type of noise. 
https://en.m.wikipedia.org/wiki/Gradient_noise 

Godot is the oddball by calling color ramps gradients.

aaronfranke replied:
Interesting. Godot 2 used the name ColorRamp, but in Godot 3 it was renamed to Gradient. If it goes against the rest of the industry then that seems like a strange change, and I can see how it's confusing in the context of Gradient noise, but without this context it seems like a good rename (Gradient sounds better to me).

Geometror replied:
As Zylann suggested, I think `color_gradient` (or in this specific case `modulation_color_gradient`) would be fine, since it sets it apart enaugh from a gradient in the context of noise generation.

TrickMe commented:
godotengine/godot-proposals #5007

TrickMe commented:
godotengine/godot-proposals#5007
